File: src/orbit/api/routes/loop.py
"""Loop 模式 API 路由。"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

router = APIRouter(prefix="/loop", tags=["loop"])


class CreateLoopRequest(BaseModel):
    # P2-5: interval 格式校验——数字+单位或 cron 5字段
    interval: str = Field(
        ...,
        min_length=2,
        max_length=50,
        pattern=r"^(\d+\s*(s|sec|second|m|min|minute|h|hr|hour|d|day)|hourly|daily|weekly|[\d\*/,]+\s+[\d\*/,]+\s+[\d\*/,]+\s+[\d\*/,]+\s+[\d\*/,]+)$",
        description="30s/5m/1h/hourly/daily/0 9 * * *",
    )
    command: str = Field(..., min_length=1)


def _get_scheduler(request: Request):
    sched = getattr(request.app.state, "loop_scheduler", None)
    if sched is None:
        raise HTTPException(status_code=503, detail="LoopScheduler 未初始化")
    return sched


@router.post("")
async def create_loop(request: Request, req: CreateLoopRequest):
    sched = _get_scheduler(request)
    loop = await sched.create(req.interval, req.command)
    await sched.start(loop.id)
    return {
        "code": 0,
        "data": {
            "loop_id": loop.id,
            "interval_seconds": loop.interval_seconds,
            "command": loop.command,
            "status": loop.status,
        },
    }


@router.get("")
async def list_loops(request: Request):
    sched = _get_scheduler(request)
    loops = sched.list_all()
    return {
        "code": 0,
        "data": {
            "loops": [
                {
                    "id": l.id,
                    "interval_seconds": l.interval_seconds,
                    "command": l.command,
                    "status": l.status,
                    "run_count": l.run_count,
                }
                for l in loops
            ],
            "total": len(loops),
        },
    }


@router.delete("/{loop_id}")
async def stop_loop(request: Request, loop_id: str):
    await _get_scheduler(request).stop(loop_id)
    return {"code": 0, "data": {"loop_id": loop_id, "status": "stopped"}}


@router.post("/{loop_id}/pause")
async def pause_loop(request: Request, loop_id: str):
    await _get_scheduler(request).pause(loop_id)
    return {"code": 0, "data": {"status": "paused"}}


@router.post("/{loop_id}/resume")
async def resume_loop(request: Request, loop_id: str):
    await _get_scheduler(request).resume(loop_id)
    return {"code": 0, "data": {"status": "active"}}

File: frontend/src/stores/loop.ts
/** Loop Store（PR4）——定时任务（Cron-like）管理。
 *
 * 对应后端 /api/v1/loop/* 路由。定时重复执行 shell 命令。
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { apiGet, apiPost, apiDelete } from '@/services/api'

export interface LoopTask {
  id: string
  interval_seconds: number
  command: string
  status: string
  run_count: number
}

export const useLoopStore = defineStore('loop', () => {
  const loops = ref<LoopTask[]>([])
  const loading = ref(false)
  const error = ref('')

  async function fetchLoops() {
    loading.value = true
    error.value = ''
    try {
      const data = await apiGet<{ loops: LoopTask[]; total: number }>('/api/v1/loop')
      loops.value = data.loops || []
    } catch {
      error.value = '加载定时任务失败'
      loops.value = []
    } finally {
      loading.value = false
    }
  }

  async function createLoop(interval: string, command: string) {
    await apiPost('/api/v1/loop', { interval, command })
    await fetchLoops()
  }

  async function stopLoop(id: string) {
    await apiDelete(`/api/v1/loop/${id}`)
    await fetchLoops()
  }

  async function pauseLoop(id: string) {
    await apiPost(`/api/v1/loop/${id}/pause`, {})
    await fetchLoops()
  }

  async function resumeLoop(id: string) {
    await apiPost(`/api/v1/loop/${id}/resume`, {})
    await fetchLoops()
  }

  return { loops, loading, error, fetchLoops, createLoop, stopLoop, pauseLoop, resumeLoop }
})
